fix browse of markdown files with line ranges

andrewhsu opened the pull request:
Fixes #4307

The first 3 commits clean up some related tests so we have thorough code coverage.

The significant patch for the issue is in f8989f4. Rewrote func ` parseFileArg` as func `genPath` because had to add logic to change  part of the url path from `/tree/` to `/blob/` for `?patch=1` to work on Markdown files.

andrewhsu committed:
fix browse command test for invalid line args

Fix tests so they trigger the error they were intending to trigger.

andrewhsu committed:
add more browse command tests to validate errors

These new tests covers code that was not previously covered.

andrewhsu committed:
remove overlapping tests for browse command

The Test_parseFileArg tests overlap code coverage of Test_runBrowse
tests.

andrewhsu committed:
fix browse of markdown files with line ranges

This fix changes the URL generated by `gh browse` so it will navigate
the user to a page that properly displays highlighted lines for Markdown files.

Changed `gh browse README.md:3-5` from generating URLs like this:
https://github.com/cli/cli/tree/trunk/README.md#L3-L5

To this:
https://github.com/cli/cli/blob/trunk/README.md?plain=1#L3-L5

soffieswan042 reviewed (commented):
#4310

mislav commented on pkg/cmd/browse/browse.go:
```diff
@@ -164,34 +166,58 @@ func runBrowse(opts *BrowseOptions) error {
-		lineRange := strings.Split(arr[1], "-")
+	filename := arr[0]
+	if len(arr) > 1 { // expecting a line number and possibly range
+		if strings.HasSuffix(filename, ".md") || strings.HasSuffix(filename, ".markdown") {
```
The following are also markdown extensions: `.mdown`, `.mkdn`. On top of those, [there are other markup formats](https://github.com/github/markup#markups) like Markdown that are being rendered by default and would need `?plain=1` when combined with line numbers:
* Textile
* RDoc
* Org-syntax
* Creole
* Mediawiki
* ReStructured text
* AscciiDoc
* Pod

So we should either try to detect all these file extensions, or give up on trying to list them all (since the list may fall out of date) and always include `?plain=1` when line numbers were specified. Which sounds better?

samcoe replied:
If there are no downsides to using `blob` and `?plain=1` then I vote we go down that path rather than maintaining a list of markdown formats.

andrewhsu replied:
Interesting, thanks for that! I did not know about the https://github.com/github/markup repo.

We could keep an exhaustive list of file extensions in this repo to match those used in `markup`, but doing so would require users of `cli` to upgrade to a new version if there is a change in behaviour to keep in sync with GitHub Web UI. I feel users would possibly perceive new file extensions rendered without line numbers as a new bug if they did not keep up with the latest version of `cli`.

Looks like `?plain=1` has no ill effect on file types that already display the line numbers, e.g. https://github.com/cli/cli/blob/trunk/go.mod?plain=1#L1-L3 looks same as https://github.com/cli/cli/blob/trunk/go.mod#L1-L3

So, always including `?plain=1` also sounds like the better option to me.

andrewhsu replied:
This PR should be ready for another review.

I've added commits adf098d and 0e0fe76 to this PR to include `?plain=1` for all files when `gh browse` is called with line numbers or line range.

Let me know if you'd prefer I squish them into f8989f4 and force-push to tidy up the git commits.

Or let me know if there are any other changes I should implement.

mislav reviewed (changes requested):
This is great! Thanks for taking this on

andrewhsu committed:
always use `?plain=1` for browse with line numbers

There are types of files other than Markdown that GitHub needs the
`?plain=1` arg to show line numbers. Therefore, just add this arg to all
URLs created from `browse` command.

andrewhsu committed:
remove overlapping tests in Test_runBrowse

The tests removed overlap code coverage for existing unit tests.
Removing so they do not need to be maintained.

mislav committed:
:nail_care: simplify `browse` implementation

- Construct URLs all at once instead of piece-by-piece
- Verify input formats before consulting the API

mislav reviewed (approved):
Thank you! I've pushed changes to minimize the nesting depth and avoid using string builder to construct URLs since I find that brittle/hard to follow.
